upb/mini_table.c

```diff
@@ -0,0 +1,937 @@
+
+  ret->size = 0;
+  ret->field_count = 0;
+  ret->ext = kUpb_ExtMode_IsMessageSet;
+  ret->dense_below = 0;
+  ret->table_mask = 0;
+  ret->required_count = 0;
+  return ret;
+}
+
+upb_MiniTable* upb_MiniTable_BuildMapEntry(upb_FieldType key_type,
+                                           upb_FieldType value_type,
+                                           upb_MiniTablePlatform platform,
+                                           upb_Arena* arena) {
+  upb_MiniTable* ret = upb_Arena_Malloc(arena, sizeof(*ret));
+  upb_MiniTable_Field* fields = upb_Arena_Malloc(arena, sizeof(*fields) * 2);
+  if (!ret || !fields) return NULL;
+
+  upb_MiniTable_Sub* subs = NULL;
+  if (value_type == kUpb_FieldType_Message || value_type == kUpb_FieldType_Group) {
+    subs = upb_Arena_Malloc(arena, sizeof(*subs));
+    if (!subs) return NULL;
+  }
+
+  size_t field_size =
+      upb_MtDecoder_SizeOfRep(kUpb_FieldRep_StringView, platform);
+
+  fields[0].number = 1;
+  fields[1].number = 2;
+  fields[0].mode = kUpb_FieldMode_Scalar;
+  fields[1].mode = kUpb_FieldMode_Scalar;
+  fields[0].presence = 0;
+  fields[1].presence = 0;
+  fields[0].descriptortype = key_type;
+  fields[1].descriptortype = value_type;
+  fields[0].offset = 0;
+  fields[1].offset = field_size;
+  fields[1].submsg_index = 0;
+
+  ret->size = UPB_ALIGN_UP(2 * field_size, 8);
+  ret->field_count = 2;
+  ret->ext = kUpb_ExtMode_NonExtendable | kUpb_ExtMode_IsMapEntry;
+  ret->dense_below = 2;
+  ret->table_mask = 0;
+  ret->required_count = 0;
+  ret->subs = subs;
+  ret->fields = fields;
+  return ret;
+}
+
+upb_MiniTable_Extension* upb_MiniTable_BuildExtensions(const char* data,
+                                                       size_t len,
+                                                       size_t* ext_count,
+                                                       upb_Arena* arena,
+                                                       upb_Status* status) {
+  upb_MtDecoder decoder = {
+      .arena = arena,
+      .status = status,
+      .table = NULL,
+  };
+
+  upb_MiniTable_Extension* exts;
+
+  if (UPB_SETJMP(decoder.err)) {
+    exts = NULL;
+    *ext_count = 0;
+    goto done;
+  }
+
+  uint16_t count = 0;
+  exts = upb_Arena_Malloc(arena, len);
+  upb_MtDecoder_CheckOutOfMemory(&decoder, exts);
+  upb_MtDecoder_Parse(&decoder, data, len, exts, sizeof(*exts), &count, NULL);
+  exts = upb_Arena_Realloc(arena, exts, sizeof(*exts) * len,
+                           sizeof(*exts) * count);
+
+done:
+  *ext_count = count;
+  return exts;
+}
+
+upb_MiniTable* upb_MiniTable_Build(const char* data, size_t len,
+                                   upb_MiniTablePlatform platform,
+                                   upb_Arena* arena, upb_Status* status) {
+  void* buf = NULL;
+  size_t size = 0;
+  upb_MiniTable* ret = upb_MiniTable_BuildWithBuf(data, len, platform, arena,
+                                                  &buf, &size, status);
+  free(buf);
```

Comment:
assert(buf != null)

Reply:
`buf` could still be NULL if the field list was empty. `free(NULL)` is ok.